api/src/main/java/marquez/api/JobResource.java

```diff
@@ -213,6 +226,28 @@ void throwIfNotExists(@NonNull RunId runId) throws MarquezServiceException {
     }
   }
 
+  private void throwIfJobDoesNotMatchRun(RunId runId, String namespaceName, String jobName) {
+    Optional<Run> runRow = runService.getRun(runId);
+    if (runRow.isEmpty()) {
+      throw new RunNotFoundException(runId);
+    }
+    Run run = runRow.get();
+    if (!jobName.equals(run.getJobName()) || !namespaceName.equals(run.getNamespaceName())) {
+      throw new RunNotFoundException(runId);
+    }
+  }
+
+  private void throwIfDatasetsNotExist(ImmutableSet<DatasetId> datasets) {
+    if (datasets == null) {
+      return;
+    }
+    for (DatasetId datasetId : datasets) {
+      if (!datasetDao.exists(datasetId.getNamespace().getValue(), datasetId.getName().getValue())) {
```

Comment:
I would add the method `DatasetService.exists(DatasetId)` to ensure we properly encapsulate the dao layer behind service calls.